rust/agents/relayer/src/merkle_tree_builder.rs

```diff
@@ -8,20 +8,63 @@ use color_eyre::eyre::Result;
 use ethers::core::types::H256;
 use std::fmt::Display;
 
-use tracing::{debug, error};
+use tracing::{debug, error, info};
+
+// Helper type to denote the fact that a Inbox that has received no checkpoint is different from an Inbox that has been checkpointed with checkpoint index 0
+#[derive(Debug, Copy, Clone)]
+pub struct OnchainCheckpointIndex {
+    index: Option<u32>,
+}
+
+impl OnchainCheckpointIndex {
+    pub fn from_checkpoint(checkpoint: &Checkpoint) -> Self {
+        Self {
+            index: if checkpoint.root.is_zero() {
+                None
+            } else {
+                Some(checkpoint.index)
+            },
+        }
+    }
+
+    pub fn from_index(index: u32) -> Self {
+        Self { index: Some(index) }
+    }
+
+    pub fn matches_signed_checkpoint(&self, other_checkpoint_index: u32) -> bool {
+        match self.index {
+            Some(index) => index >= other_checkpoint_index,
+            None => false,
+        }
+    }
+
+    pub fn is_behind_prover(&self, prover_count: usize) -> bool {
```

Comment:
This also feels a little confusing for similar reasons as `matches_signed_checkpoint`, I'm not sure this impl should be aware of the prover count

Just thinking of a more generalizable way of doing comparisons-- I believe you can implement [PartialOrd](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) for types other than Self which could be cool, maybe instead of `matches_signed_checkpoint` and `is_behind_prover` you could just implement `impl PartialOrd<u32> for OnchainCheckpointIndex`?